// helpers/modules/logger/index.js
const fs = require('fs');
const path = require('path')

const { timestamp } = require('helpers/modules/time');
const { logTypeArgument } = require('helpers/modules/disambiguate');

function addLogToFile(logObject) {
    const time = timestamp.time();
    const date = timestamp.date();

    const sanitizedLogType = logTypeArgument(logObject.logType);
    const logmessage = `(${time}) ${sanitizedLogType}:${logObject.message}\n`;

    try {
        const resolveFilePath = path.resolve(__dirname, `store/${date}.log`);
        fs.appendFileSync(resolveFilePath, logmessage);
    } catch (error) {
        displayLog({"message": error, "logType": 'E'});
    }
}

function displayLog(logObject) {
    const sanitizedLogType = logTypeArgument(logObject.logType);
    const time = timestamp.time();

    console.log(`(${time}) ${sanitizedLogType}:${logObject.message}`);
}

function log(logObject) {
    displayLog(logObject);
    addLogToFile(logObject);
}

module.exports = {
    logger: {
        log: log,
        display: displayLog,
        write: addLogToFile
    }
};
