// helpers/modules/time/index.js
function getTime() {
    const currentDate = new Date();

    const hour_24 = currentDate.getHours();     //current hour in 24 hour format
    const hour_12 = formatHour(hour_24);        //current hour in 12 hour format + AM/PM suffix
    const minutes = currentDate.getMinutes();
    const seconds = currentDate.getSeconds();

    const time = {
        hour_24: hour_24,
        hour_12: hour_12,
        minutes: minutes,
        seconds: seconds
    };

    return time;
}

function getDate() {
    const currentDate = new Date();

    const day = currentDate.getDay();
    const dayName = getDayName(day);

    const dayOfMonth = currentDate.getDate();

    const monthNumber = currentDate.getMonth();
    const monthName = getMonthName(monthNumber);

    const year = currentDate.getFullYear();

    const date = {
        day_name: dayName,
        day_of_month: dayOfMonth,
        month_number: monthNumber + 1,
        month_name: monthName,
        year: year
    };

    return date;
}

function getMonthName(monthNumber) {
    const monthNames = require('helpers/modules/time/config/months.json');

    return monthNames[monthNumber];
}

function getDayName(dayNumber) {
    const dayNames = require('helpers/modules/time/config/day-names.json');

    return dayNames[dayNumber];
}

function formatHour(hour) {
    const hourSuffix = hour > 11 ? `PM` : 'AM';
    const formattedHour = hour > 12 ? hour - 12 : hour;

    const hour_12 = {
        hour: formattedHour,
        suffix: hourSuffix
    };

    return hour_12;
}

module.exports = {
    timestamp: {
        time: getTime,
        date: getDate
    }
};
